src/consumer/stream_consumer.rs

```diff
@@ -49,76 +48,86 @@ unsafe fn disable_nonempty_callback(queue: &NativeQueue) {
     rdsys::rd_kafka_queue_cb_event_enable(queue.ptr(), None, ptr::null_mut())
 }
 
-struct WakerSlab {
-    wakers: Mutex<Slab<Option<Waker>>>,
+/// Data structure to store waker instances from idle streams.
+/// This queue wakes up the wakers in FIFO order.
+/// On registration streams receive an id they can reuse for setting and removing wakers.
+///
+/// Implementation is optimized for adding new wakers and waking all of them.
+/// Removing a waker is less efficient.
+struct WakerQueue {
+    /// Vec containing the set waker instances and the ids of their owning streams
+    waker_queue: Mutex<Vec<(Uuid, Waker)>>,
 }
 
-impl WakerSlab {
-    fn new() -> WakerSlab {
-        WakerSlab {
-            wakers: Mutex::new(Slab::new()),
+impl WakerQueue {
+    fn new() -> WakerQueue {
+        WakerQueue {
+            waker_queue: Mutex::new(Vec::new()),
         }
     }
 
+    /// Awakes all set wakers in order of their arrival
     fn wake_all(&self) {
-        let mut wakers = self.wakers.lock().unwrap();
-        for (_, waker) in wakers.iter_mut() {
-            if let Some(waker) = waker.take() {
-                waker.wake();
-            }
+        let waker_queue = std::mem::take(&mut *self.waker_queue.lock().unwrap());
+        for (_, waker) in waker_queue.into_iter() {
+            waker.wake();
         }
     }
 
-    fn register(&self) -> usize {
-        let mut wakers = self.wakers.lock().expect("lock poisoned");
-        wakers.insert(None)
```

Comment:
I noticed this isn't in the project ATM. Does this mean you're also syncing the fork? If so, could you mention that explicitly?

Reply:
I don't get your point. Once this PR is merged, mp-rs should use the latest version.

Reply:
I just meant to ask if you were pulling the latest from https://github.com/fede1024/rust-rdkafka/ because some of this logic doesn't seem to exist in the DD fork of the repo ATM but the diff doesn't exactly show this 😕.

Reply:
The main branch for the dd fork is `datadog/main` and that's the branch the diff is comparing with. Don't know up-to-date it is with the original project.

Reply:
ah that's way, cool, I didn't notice. It defaults to master so I was quite confused lol